File: src/clients/mix-server-client/src/app/services/repositories/models/device.ts
import {ClientType, DeviceType} from "../../../generated-clients/mix-server-clients";
import {FileExplorerFileNode} from "../../../main-content/file-explorer/models/file-explorer-file-node";
import {DeviceState} from "./device-state";

export class Device {
  constructor(public id: string,
              public clientType: ClientType,
              public deviceType: DeviceType,
              public lastSeen: Date,
              public interactedWith: boolean,
              public isCurrentDevice: boolean,
              public online: boolean,
              public capabilities: { [mimeType: string]: boolean },
              public brand: string | undefined,
              public browserName: string | undefined,
              public model: string | undefined,
              public osName: string | undefined,
              public osVersion: string | undefined) {
  }

  public get displayName(): string {
    return `${this.brand ?? ''} ${this.model ?? ''} ${this.osName ?? ''} ${this.osVersion ?? ''} ${this.browserName ?? ''}`
  }

  public get icon(): string {
    return 'devices'
  }

  canPlay(file?: FileExplorerFileNode | null): boolean {
    return !!file && (file.hasCompletedTranscode || (this.capabilities[file.metadata.mimeType] ?? false));
  }

  updateFromState(state: DeviceState, currentUserId: string | null | undefined) {
    this.interactedWith = state.interactedWith && state.lastInteractedWith === currentUserId;
    this.online = state.online && state.lastInteractedWith === currentUserId;
    this.capabilities = state.capabilities;
  }
}

File: src/api/MixServer.Domain/Users/Entities/Device.cs
using System.Collections.Concurrent;
using System.ComponentModel.DataAnnotations.Schema;
using MixServer.Domain.FileExplorer.Models;
using MixServer.Domain.Users.Enums;
using MixServer.Domain.Users.Models;

namespace MixServer.Domain.Users.Entities;

public interface IDevice : IDeviceInfo, IDeviceState
{
    Guid Id { get; }
    DateTime LastSeen { get; }
}

public class Device : IDevice
{
    public Guid Id { get; set; }
    public DateTime LastSeen { get; set; }

    public List<UserCredential> UserCredentials { get; set; } = [];
    
    [NotMapped]
    public Guid DeviceId => Id;

    [NotMapped]
    public string LastInteractedWith { get; private set; } = string.Empty;

    [NotMapped]
    public bool InteractedWith { get; private set; }

    [NotMapped]
    public bool Online { get; set; }

    [NotMapped]
    public ConcurrentDictionary<string, bool> Capabilities { get; private set; } = new();

    public bool GetMimeTypeSupported(string? mimeType)
    {
        if (string.IsNullOrWhiteSpace(mimeType))
        {
            return false;
        }
        
        return Capabilities.TryGetValue(mimeType, out var supported) && supported;
    }

    public void Populate(IDeviceState state)
    {
        Online = state.Online;
        LastInteractedWith = state.LastInteractedWith;
        InteractedWith = state.InteractedWith;
        Capabilities = state.Capabilities;
    }

    #region DeviceInfo

    public ClientType ClientType { get; set; }
    public DeviceType DeviceType { get; set; }
    public string? BrowserName { get; set; }
    public string? Model { get; set; }
    public string? Brand { get; set; }
    public string? OsName { get; set; }
    public string? OsVersion { get; set; }

    #endregion

    public void UpdateDeviceInfo(IDeviceInfo deviceInfo)
    {
        ClientType = deviceInfo.ClientType;
        DeviceType = deviceInfo.DeviceType;
        BrowserName = deviceInfo.BrowserName;
        Model = deviceInfo.Model;
        Brand = deviceInfo.Brand;
        OsName = deviceInfo.OsName;
        OsVersion = deviceInfo.OsVersion;
    }
}